Add CMEPS to ufs-s2s-model

DeniseWorthen opened the pull request:
This PR adds a second mediator ( CMEPS ) to the ufs-s2s-model:

- Adds CMEPS version regression tests for all current NEMS mediator tests (rt.sh). All current NEMS mediator baselines were tested and found identical except for a global attribute difference in the MOM6 restarts (commit 45f515ebdd in this branch). 

- Adds a restart reproducibility test for CMEPS. A 3 day C96 baseline is created. The 2d C96 is restarted and run for an additional day. Answers are then compared against the 3 day baseline. An rt.conf.restart file is added for convenience since no baseline is created for the restart test.

- Updates namelists and nems.configure files for the coupled model to accommodate warm-start restarting of the coupled model

- Updates the namelist settings for the restart output directory for MOM6 from _MOM6_RESTART_ to _RESTART_ and the output directory for CICE5 from _restart_ to _RESTART_. Both are consistent with FV3

- Adds two additional configurable settings to MOM_input_template. Default values are consistent with current settings used in ufs-s2s-model. Restart reproducibility requires non-default values until the halo decomposition bug seen when _USE_LA_LI2016 = true_ is fixed.

- Updates NEMS to the current develop which contains a fix for CMEPS build for rt.sh and also adds required component.mk for CICE6.

This PR will become active after the associated CICE5 PR: #89 

Associated project: [2](https://github.com/ufs-community/ufs-s2s-model/projects/2)
Associated issues: #48, #32, #33, #76 

Co-author: @mvertens 

The following figures show the difference in the ocean surface state and the ice surface state after 10d for a single initial condition date for the NEMS and CMEPS mediators.

![nems_long-cmeps_long_10](https://user-images.githubusercontent.com/40498404/83429115-903c1280-a401-11ea-8a2d-ae6e766c37a2.jpg)

![aice_h_11](https://user-images.githubusercontent.com/40498404/83429134-9a5e1100-a401-11ea-8875-c5702518964c.png)
![Tsfc_h_11](https://user-images.githubusercontent.com/40498404/83429150-9df19800-a401-11ea-9755-262a8846ed2b.jpg)

The following figures show the difference in the ocean surface state after 10d for a single initial condition date when wave coupling is included.

![nems_ww3-cmeps_ww3_10](https://user-images.githubusercontent.com/40498404/83429178-aa75f080-a401-11ea-9089-39fddbe31e78.jpg)

DeniseWorthen committed:
Merge pull request #1 from ufs-community/develop

Feature/add comparisons (#4)

DeniseWorthen committed:
Update .gitmodules

DeniseWorthen committed:
update submodules for nems,stochastic_physics

DeniseWorthen committed:
update NEMS,stochastic_physics

DeniseWorthen committed:
Merge pull request #2 from ufs-community/develop

merge develop/ufs-s2s-model

DeniseWorthen committed:
update CICE submode for feature/halfdeg (with build change) and
MOM_input template for initial condition files

DeniseWorthen committed:
Merge pull request #3 from ufs-community/develop

merge fv3atm/develop

DeniseWorthen committed:
Revert "update CICE submode for feature/halfdeg (with build change) and"

This reverts commit 0c8cb0b8451a11f7ab897cb4d31fc6233cf96c31.

DeniseWorthen committed:
update CICE submodule

DeniseWorthen committed:
Add CMEPS submodule

DeniseWorthen committed:
update submodules for feature/cmeps_integration branches of FV3
and NEMS

DeniseWorthen committed:
update conf,modulefiles and parm for CMEPS
add appbuilder files for CMEPS

DeniseWorthen committed:
add compsets for CMEPS

DeniseWorthen committed:
fix cmeps compsets for 2d warm

DeniseWorthen committed:
update develop branch in my fork to use NOAA-EMC develop branches

Commit:
udpated NEMS to remove CMEPS as much as possible

Commit:
updates to CMEPS

Commit:
updated NEMS

Commit:
updated NEMS

Commit:
updated NEMS

Commit:
updated FV3

DeniseWorthen committed:
update NEMS submodule

DeniseWorthen committed:
change coldair_outbreak_mod to false in cold start to match warm
cmeps nems.configure

DeniseWorthen committed:
Merge pull request #5 from ufs-community/develop

merge to latest develop

DeniseWorthen committed:
Merge pull request #6 from ufs-community/develop

Debug mode capablility (#15)

DeniseWorthen committed:
Update configure.fv3_coupled.hera.intel

DeniseWorthen committed:
Update configure.fv3_coupled.cheyenne.intel

DeniseWorthen committed:
update submodules

DeniseWorthen committed:
Merge pull request #7 from ufs-community/develop

Add debug compilation flag specification at top level (#17)

DeniseWorthen committed:
Merge pull request #8 from ufs-community/develop

update to current ufs-weather-app (#19)

DeniseWorthen committed:
update gitmodules, FV3 for gfsv16 incomming commit, NEMS for
nomerg2atm, new MOM6 dev/emc

DeniseWorthen committed:
Revert changes that were meant for a branch, not develop

This reverts commit 57236a665d1efcbac28a53826daaa7a0385746eb.

DeniseWorthen committed:
update submodules for new MOM6 commit, FV3 gfsv16
commit w/ min,max on ice_fraction

DeniseWorthen committed:
update conf,modulefiles,parm consistent w/
update of ufs-s2s-model

DeniseWorthen committed:
update NEMS submodule, rename debug appbuilder for NEMS case

DeniseWorthen committed:
Merge pull request #9 from ufs-community/develop

Update FV3, NEMS (#21)

DeniseWorthen committed:
update NEMS

DeniseWorthen committed:
Merge pull request #13 from ufs-community/develop

Update2020 01 (#23)

DeniseWorthen committed:
add compsets for running cmeps in datesuite mode

DeniseWorthen committed:
changes with latest cmeps, field name change branches

DeniseWorthen committed:
update FV3 submodule

DeniseWorthen committed:
update NEMS

DeniseWorthen committed:
update CMEPS submodule; update configure and modulefiles
for esmf8.1.0bs10 on both cheyenne and hera

DeniseWorthen committed:
Merge branch 'mvertens/cmeps_integration_update' of https://github.com/DeniseWorthen/ufs-s2s-model into mvertens/cmeps_integration_update

DeniseWorthen committed:
update module files for testing 8.1bs10 on cheyenne

DeniseWorthen committed:
Merge pull request #15 from ufs-community/develop

DeniseWorthen committed:
update to develop for submodules CICE,NEMS. CMEPS points to branch
dworthen/nems_integration2_update2 which no longer contains changes
for esmf8.1.0bs10. Tony's additional logging and meshcreated logic
is included.

DeniseWorthen committed:
Merge pull request #17 from ufs-community/develop

Update NEMS,FV3,CICE (#28)

DeniseWorthen committed:
update conf and modulefiles

DeniseWorthen committed:
add fingerprint for updated baseline area

DeniseWorthen committed:
update CMEPS submodue for memory leak fix

DeniseWorthen committed:
add cold,1d,2d,restart compsets for cmeps

DeniseWorthen committed:
fixed typo in baseline name; comparisons currently fail in tests for
mediator restart files but manual comparisons using nccmp -d show
no differences

DeniseWorthen committed:
update FV3,NEMS for branches containing field name changes
update CMEPS to master
update compsets & parm consistent w/ rts created for cmeps

DeniseWorthen committed:
Updates to MOM6 cap

DeniseWorthen committed:
update submodules for branches containing timestamps
on filed dumps;also mom6 for restartcontrol and cice5 for
restart repro

DeniseWorthen committed:
update MOM6

DeniseWorthen committed:
remove use of MOM_input NAME_ROOT which will not be
required w/ recent changes in mom_cap for restart
fixes

DeniseWorthen committed:
update CMEPS,MOM6 submodules; fixes for compsets; adds
fv3 restart interval used in model_configure

DeniseWorthen committed:
update MOM6

DeniseWorthen committed:
update MOM6 w/ fix for last output file not being written

DeniseWorthen committed:
update CMEPS (updated yaml, comment out histavg_n)
FV3 points to Dom's b4b branch w/ timestamps on
field dumps and field name changes for CMEPS
MOM6 with Mariana's cesm-cleanup

DeniseWorthen committed:
update MOM6 w/ changes to mct driver to match dev/ncar

DeniseWorthen committed:
update CICE submodule

DeniseWorthen committed:
Merge pull request #18 from ufs-community/develop

DeniseWorthen committed:
update CMEPS, MOM6

DeniseWorthen committed:
update NEMS

DeniseWorthen committed:
update compsets and parm to bring cmeps into s2s

DeniseWorthen committed:
two more cmeps related changes

DeniseWorthen committed:
update FV3, NEMS, compsets, conf, modulefiles and
parm to get closer to ufs-s2s after ccpp was added

DeniseWorthen committed:
Merge pull request #19 from ufs-community/develop

Add CCPP to UFS-S2S-model

DeniseWorthen committed:
update FV3

DeniseWorthen committed:
switched to new FV3 branch w/ correct ccpp submods
updated FMS

DeniseWorthen committed:
updates for compsets using testsize

DeniseWorthen committed:
update compsets

DeniseWorthen committed:
update compsets for trying to use testsize and fixes
for restart compsets using fhrot

DeniseWorthen committed:
reverted to batch queues; all tests pass for nems; restart
mediator file failes for cmeps in RT but passes if use nccmp -d

DeniseWorthen committed:
update to punyswrad branch of CICE

DeniseWorthen committed:
update 2 parm files for reproducibility; update compsets

DeniseWorthen committed:
give extra time for comparisons

DeniseWorthen committed:
update MOM6

DeniseWorthen committed:
update MOM6 for restart_mode = nems

DeniseWorthen committed:
update MOM6

DeniseWorthen committed:
update MOM6

DeniseWorthen committed:
update MOM6 after log cleanup; comment out sfc_tile
comparisons

DeniseWorthen committed:
begin modifying ufs-s2s w/ optional ccpp to be default
test but with either nems or cmeps

DeniseWorthen committed:
updates for compsets

DeniseWorthen committed:
more incremental compset changes

DeniseWorthen committed:
update FV3 for tsfl fix for restart repro testing

DeniseWorthen committed:
Merge branch 'feature/addccpp' of https://github.com/DeniseWorthen/ufs-s2s-model into feature/addccpp

DeniseWorthen committed:
Merge pull request #20 from ufs-community/develop

DeniseWorthen committed:
update gitmodules to point to master at ESCOMP for
CMEPS and dev/emc for MOM6; update FV3 to latest
develop w/ changes for fieldnames

DeniseWorthen committed:
update submodules

DeniseWorthen committed:
add WW3 as component and associated parm files. Update
FV3 for ccpp/physics tsfcl fix

DeniseWorthen committed:
update FV3,NEMS submodules

DeniseWorthen committed:
update FV3

DeniseWorthen committed:
Merge pull request #22 from ufs-community/develop

Add debug (#70)

DeniseWorthen committed:
Merge pull request #23 from ufs-community/develop

DeniseWorthen committed:
start building alt compset run version of cmeps/nems
ccpp prod only

DeniseWorthen committed:
changes consistent w/ Minsuk's altcompset branch

DeniseWorthen committed:
add tests; ccpp prod only

DeniseWorthen committed:
modifications for cmeps

DeniseWorthen committed:
Merge pull request #24 from ufs-community/develop

Remove IPD tests and CCPP repro tests from the regression test system…

DeniseWorthen committed:
updating to current s2s where applicable

DeniseWorthen committed:
updating compsets

DeniseWorthen committed:
Merge pull request #25 from ufs-community/develop

Feature/alt compsetrun (#75)

DeniseWorthen committed:
update MOM6 and CICE5 submodules for porting to
orion

DeniseWorthen committed:
building tests for cmeps

DeniseWorthen committed:
Merge pull request #26 from ufs-community/develop

Porting regression tests to Orion (#78)

DeniseWorthen committed:
update to match current ufs-s2s w/ rt.sh tests

DeniseWorthen committed:
Merge branch 'develop' into feature/cmeps_altrt

DeniseWorthen committed:
Merge pull request #27 from ufs-community/develop

Update NEMS submodule

DeniseWorthen committed:
Merge pull request #28 from DeniseWorthen/develop

DeniseWorthen committed:
fix missing quote in all cmeps tests

DeniseWorthen committed:
add temporart rt.conf for testing

DeniseWorthen committed:
Merge pull request #31 from ufs-community/develop

wave-ocean coupling (#91)

DeniseWorthen committed:
update for current develop of S2S after wave ocean coupling
merge

DeniseWorthen committed:
update submodules; cice/cmeps for rhoa changes

DeniseWorthen committed:
fixes for tests

DeniseWorthen committed:
more fixes

DeniseWorthen committed:
more tests fixes; changed case name for cmeps and fixed
failure to copy mediator restart file; cold tests complete
for cmeps

DeniseWorthen committed:
Delete fv3_mom6_cice_0.exe

DeniseWorthen committed:
Delete modules.fv3_mom6_cice_0

DeniseWorthen committed:
more tests fixes for cmeps

DeniseWorthen committed:
add back the cmeps-ipd app builder file for testing

DeniseWorthen committed:
update CICE and CMEPS submodules

DeniseWorthen committed:
update NEMS for compile.sh fix for cmeps

DeniseWorthen committed:
Merge pull request #33 from ufs-community/develop

update cice5 (#89)

DeniseWorthen committed:
update CMEPS for yaml changes

DeniseWorthen committed:
add cmeps cold debug test; add settings to use 1d bmark
for restart test

DeniseWorthen committed:
update CMEPS

DeniseWorthen committed:
fix cmeps cold sequence with waves

DeniseWorthen committed:
update CMEPS

DeniseWorthen committed:
Point back to escomp/cmeps after merge of ww3 yaml changes
Add proto-restart test (not yet working); Add intermediate
restart writing during 1d benchmark RT for CMEPS. Add
temporary rt.conf for cmeps versions

DeniseWorthen committed:
final updates prior to merge with ufs-s2s@develop

DeniseWorthen committed:
Merge branch 'develop' into feature/cmeps_altrt

DeniseWorthen committed:
Initial regression testing of this branch using NEMS mediator
after updating to current ufs-s2s@develop
No changes to answers are expected
Tests fail only because of the global attribute in MOM6
MOM6 restart files

DeniseWorthen committed:
remove rpointer.cpl from comparison list and instead
create it using the date strings; add fixes for restart
test

DeniseWorthen committed:
remove pnetcdf module and change pnetcdf to netcdf
in med_modelio.nml. add cmeps 6h debug test w/ 1hour
wall clock

DeniseWorthen committed:
final commit of restart test

DeniseWorthen committed:
Merge pull request #34 from ufs-community/develop

Orion post maintenance fix (#102)

DeniseWorthen committed:
Merge branch 'develop' into feature/cmeps_altrt

DeniseWorthen committed:
commit combined rt.conf with both nems and cmeps tests

DeniseWorthen committed:
Delete rt.conf.cmeps

DeniseWorthen committed:
add C96 3d test for use with a C96 restart test. Add additional
run file to setup C96 restart test

DeniseWorthen committed:
update rt.sh with new baseline area

DeniseWorthen commented:
Regression tests are currently running on Hera.

DeniseWorthen committed:
add 3d test and restart test to rt.conf for verifying baselines

JessicaMeixner-NOAA commented on conf/configure.fv3_coupled.gaea.intel:
```diff
@@ -94,7 +94,7 @@ endif
 FFLAGS_OPT = -O2 -debug minimal -fp-model source -qoverride-limits -qopt-prefetch=3
 FFLAGS_REPRO = -O2 -debug minimal -fp-model consistent -qoverride-limits -g -traceback
-FFLAGS_DEBUG = -g -O0 -check all -check noarg_temp_created -check nopointer -warn -warn noerrors -fp-stack-check -fstack-protector-all -fpe0 -debug -traceback -ftrapuv
+FFLAGS_DEBUG = -g -O0 -check all -check noarg_temp_created -check nopointer -warn -warn noerrors -fp-stack-check -fstack-protector-all -fpe0 -debug -traceback -ftrapuv -link_mpi=dbg        
```
Should we also add 
ifeq ($(CMEPS),Y)
 CPPDEFS += -DCMEPS
 endif
to this an the other fv3_coupled configure files? (particularly wcoss-dell at the least?)

JessicaMeixner-NOAA commented on parm/nems.configure.medcmeps_atm_ocn_ice.IN:
```diff
@@ -0,0 +1,98 @@
+      mediator_read_restart = .true.
+::
+MED_attributes::
+      ATM_model = fv3gfs
```
Maybe it's just me, but I think it's kind of awkward that these are different names than what is above for ATM_model and OCN_model here than it is above when we define these?  mom versus mom6?

DeniseWorthen replied:
I believe that the MED_attributes here are used only distinguish between having a active component and having a stub. So what is checked by CMEPS is that the ATM_model is not 'satm' or the OCN_model is not 'socn'.

JessicaMeixner-NOAA replied:
If that's the case, can we make this consistent with the top of the file and use the @[atm_model], @[ocn_model] etc that is used at the top of this file?

DeniseWorthen replied:
I've tested this and it works so I will change it.

JessicaMeixner-NOAA commented on parm/nems.configure.medcmeps_atm_ocn_ice_wav.IN:
```diff
@@ -0,0 +1,112 @@
+
+# CMEPS variables
+
+component_list: ATM OCN ICE MED
```
Also this component list is different than the component list above.  How do we know this component list is for the mediator versus the EARTH_component_list above?

JessicaMeixner-NOAA replied:
Thanks for making these changes!

JessicaMeixner-NOAA commented on tests/rt.conf.restart:
```diff
@@ -0,0 +1,3 @@
+COMPILE | coupledFV3_CCPP_MOM6_CICE_CMEPS           |   | hera.intel  | fv3 |
```
Since this is not part of the standard rt.conf, how do we run this test? How does this get tested and then the log file for reporting not be overwritten?

DeniseWorthen replied:
The restart test is awkward. Since it doesn't generate a baseline, it is only run against current or new baselines. And the 3d baseline must have already been created. It would be good if there could be a conditional in rt.sh---run the restart test only after running the 2d and 3d tests and not if you're creating baselines. 

I have this rt.conf.restart just as a convenience so that I can cut/paste the restart run line into the rt.conf for the final comparison. I could remove the file or rename it.

DeniseWorthen replied:
I will remove this file.

JessicaMeixner-NOAA reviewed (commented):
Just a few comments/questions, but overall looks good to me.  I'll hit the approve after the regtests all go through.

DeniseWorthen committed:
add  Hera logs

DeniseWorthen committed:
Remove component_list from nems.configure files since it is
only used by CIME driver.
Add ifdef CMEPS to fv3_coupled for all platforms.
Re-order rt.conf so that when executed sequentially, all cold
jobs are done first.
For restart test, set up to copy restarts from the 'dependent run'
directory.

DeniseWorthen committed:
add Orion logs

DeniseWorthen committed:
Add Hera logs

Reran tests w/ changes to nems.configure and
adding DEP_RUN variable (currently manually set) in restart test

JessicaMeixner-NOAA reviewed (approved):
This is great!

binli2337 reviewed (approved):
Looks good.

junwang-noaa reviewed (approved):
Thanks for making changes in restart test.  Please remove the tests/output, it seems a temporary file to me.
We may need to discuss on how to start the model later.

DeniseWorthen committed:
Delete output

DeniseWorthen commented:
Thanks Everyone for your timely reviews.
